include/libdnf5/base/transaction_package.hpp

```diff
@@ -43,70 +42,60 @@ class TransactionPackage {
     //
     // @replaces libdnf:transaction/TransactionItem.hpp:method:TransactionItemBase.getReason()
-    Reason get_reason() const noexcept { return reason; }
+    Reason get_reason() const noexcept;
 
     /// @return packages replaced by this transaction package.
-    const std::vector<rpm::Package> get_replaces() const noexcept { return replaces; }
+    std::vector<rpm::Package> get_replaces() const noexcept;
 
     /// @return packages that replace this transaction package (for transaction
     /// packages that are leaving the system).
-    const std::vector<rpm::Package> & get_replaced_by() const noexcept { return replaced_by; }
+    const std::vector<rpm::Package> & get_replaced_by() const noexcept;
 
     /// The REASON_CHANGE action requires group id in case the reason is changed to GROUP
     /// @return id of group the package belongs to
-    const std::string * get_reason_change_group_id() const noexcept {
-        return reason_change_group_id ? &reason_change_group_id.value() : nullptr;
-    }
+    const std::string * get_reason_change_group_id() const noexcept;
+
+    ~TransactionPackage();
+
+    TransactionPackage(const TransactionPackage & mpkg);
+    TransactionPackage & operator=(const TransactionPackage & mpkg);
+    TransactionPackage(TransactionPackage && mpkg) noexcept;
+    TransactionPackage & operator=(TransactionPackage && mpkg) noexcept;
 
 private:
     friend class Transaction::Impl;
     friend class ::BaseGoalTest;
     friend class ::RpmTransactionTest;
 
-    TransactionPackage(const libdnf5::rpm::Package & pkg, Action action, Reason reason)
-        : package(pkg),
-          action(action),
-          reason(reason) {}
+    TransactionPackage(const libdnf5::rpm::Package & pkg, Action action, Reason reason);
 
-    TransactionPackage(const libdnf5::rpm::Package & pkg, Action action, Reason reason, State state)
-        : package(pkg),
-          action(action),
-          reason(reason),
-          state(state) {}
+    TransactionPackage(const libdnf5::rpm::Package & pkg, Action action, Reason reason, State state);
 
     TransactionPackage(
-        const libdnf5::rpm::Package & pkg, Action action, Reason reason, std::optional<std::string> group_id)
-        : package(pkg),
-          action(action),
-          reason(reason),
-          reason_change_group_id(group_id) {}
-
-    void set_reason(Reason value) noexcept { reason = value; }
-    void set_state(State value) noexcept { state = value; }
-
-    libdnf5::rpm::Package package;
-    Action action;
-    Reason reason;
-    State state{State::STARTED};
-    std::optional<std::string> reason_change_group_id;
-
-    std::vector<rpm::Package> replaces;
-    std::vector<rpm::Package> replaced_by;
+        const libdnf5::rpm::Package & pkg, Action action, Reason reason, const std::optional<std::string> & group_id);
+
+    void set_reason(Reason value) noexcept;
+    void set_state(State value) noexcept;
+    void replaced_by_append(rpm::Package && pkg);
```

Comment:
This (and the one below) is also used only in impl right? Probably not needed in the public API.

Reply:
It will require a new private header but I can do that.

Reply:
I have moved also the `set_reason(...)` to the Impl.